edge-hub/src/Microsoft.Azure.Devices.Routing.Core/Evaluator.cs

```diff
@@ -53,41 +51,36 @@ public ISet<Route> Routes
             }
         }
 
-        public ISet<Endpoint> Evaluate(IMessage message)
-        {
-            var endpoints = new HashSet<Endpoint>();
-
-            // Because we are only reading here, it doesn't matter that it is under a lock
-            // ReSharper disable once InconsistentlySynchronizedField - compiledRoutes is immutable
-            ImmutableDictionary<string, CompiledRoute> snapshot = this.compiledRoutes;
-            foreach (CompiledRoute compiledRoute in snapshot.Values.Where(cr => cr.Route.Source.Match(message.MessageSource)))
-            {
-                if (EvaluateInternal(compiledRoute, message))
-                {
-                    endpoints.UnionWith(compiledRoute.Route.Endpoints);
-                }
-            }
+        public ISet<RouteResult> Evaluate(IMessage message)
+        {
+            // Because we are only reading here, it doesn't matter that it is under a lock
+            // ReSharper disable once InconsistentlySynchronizedField - compiledRoutes is immutable
+            ImmutableDictionary<string, CompiledRoute> snapshot = this.compiledRoutes;
+
+            // Multiple routes for the same endpoint can exist, in which case
+            // the message should be routed with the highest available priority
+            var routes = snapshot.Values
+                .Where(cr => cr.Route.Source.Match(message.MessageSource) && EvaluateInternal(cr, message))
+                .Select(cr => new RouteResult(cr.Route.Endpoint, cr.Route.Priority, cr.Route.TimeToLiveSecs))
+                .GroupBy(r => r.Endpoint)
+                .Select(dupes => dupes.OrderBy(r => r.Priority).First());
+
+            var results = new HashSet<RouteResult>(routes);
```

Comment:
nit: you can combine line 62 and 68 by calling ToHashSet().

Reply:
not sure if you saw my other comment, but there is no ToHashSet() that I could find, only ToImmutableHashSet(), which doesn't work here because we need to modify the collection right below.

---
In reply to: [377443584](https://github.com/Azure/iotedge/pull/2469#discussion_r377443584) [](ancestors = 377443584)